### DISCUSSION TITLE
Simplify media player entities which support  setting volume

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#1012] [Entity Models] [answered]
## Proposal

Media player entities which support setting an absolute volume should no longer implement the `volume_up` or `volume_down` services. Instead, the base entity's default implementations of the `volume_up` or `volume_down` services which call the `set_volume` service with a negative or positive adjustment will be used.

To still allow `volume_down` and `volume_up` to match the native volume up/down implementation of a device, a new property `volume_step_size` is added. If `volume_step_size` is not set, the default step size is the same as without this change (0.1).

## Rationale

There's an idea to allow setting a maximum volume for media players. This is not possible of the media player implements custom `volume_up` or `volume_down` services which change the volume by an undefined amount.

## Impact
32 core integrations with a media player platform has a custom implementation of volume_up. This is about a 3rd of the integrations with media player platforms (85).

## Problems

Subsequent calls to the default `volume_down` or `volume_up` service handlers do not work as expected if the volume level was not updated in between the calls.
For example, if the media player's volume is 0.5, a call to `volume_up` immediately followed by a call to `volume_down` will first attempt to set the volume to 0.5 + 0.1 = 0.6 and then to 0.5 - 0.1 = 0.4.
However, since this problem already exists in the default service handlers, I think this issue can be solved outside of this architecture proposal.

**Comment by mvdwetering** [comment]
I think there is a lot of value in having `volume_down` and `volume_up` match the native volume up/down implementation of a device. It allows a user to update the volume in steps in the same way as using volume up/down using a remote control or volume up/down controls in a native app would.

Making the stepsize fixed at 0.1 takes that relation away. Now the volume steps from HA will not match the stepsizes a user is expecting or used to.


Next to that volume steps of 0.1 are huge (=10% of the range). This is not a suitable range to tune the volume a bit louder/quieter. 
This of course depends on the amount of volume steps supported in the volume range by the actual device. If a device supports only 10 discrete volume steps then 10% steps are perfect. But for devices with a large volume range like an AV receiver then steps with a size of 10% of the range are just too big to use `volume_down` and `volume_up` to get to a suitable volume level.

I don't think there will be a good one-size-fits-all default volume up/down step size for all integrations.

Maybe an integration should be able to indicate the volume step size that the `volume_down` and `volume_up` calls should be?
That way an integration can set a suitable volume stepsize for the specific device/integration or not define it and fallback to a default value (which could be 0.1)

**Reply by emontnemery** [reply]
This is a good point, in a discussion with @MartinHjelmare we came to the same exact conclusion 👍 

The proposal has been updated with a new property, `volume_step_size`.

**Comment by emontnemery** [comment]
As pointed out by @drafteed in https://github.com/home-assistant/core/pull/105952, A default implementation of `volume_up` / `volume_down` which call `set_volume` will not work reliably with any polled integration if the media player volume is changed locally on the device.
As an example, Home Assistant believes the volume to be `0.5` based on the last poll. The user decreases the volume with the device's remote control to `0.1`. If the `volume_down` service is called within the same poll cycle, Home Assistant will set the volume to `0.5 - volume_step`, i.e. most likely an increased volume.

A workaround could be to call `async_update` before setting the volume.

**Reply by Drafteed** [reply]
> A workaround could be to call async_update before setting the volume.

This will cause an annoying delay before the command is executed.

**Reply by emontnemery** [reply]
Sure, there will be a delay.

When are the `volume_up` / `volume_down` services called for media players which support `set_volume` though?
We for example don't expose those services for media players which do support `set_volume` in frontend, we instead show a volume slider.

**Reply by Drafteed** [reply]
> When are the volume_up / volume_down services called for media players which support set_volume though?

1. The `+` and `-` buttons are displayed in the media player dialog. I use them sometimes.
2. [HomeKit Bridge](https://www.home-assistant.io/integrations/homekit/) controls the volume of the media player from iOS Remote Widget through `volume_up` / `volume_down` services.
3. Some voice assistants also use volume control through `volume_up` / `volume_down` services, if they receive a command like «Turn up the volume on TV» or «Make TV louder».

**Reply by mvdwetering** [reply]
>When are the volume_up / volume_down services called for media players which support set_volume though?

There are physical remotes with hardware volume up/down buttons that integrate with Home Assistant. 
Remote Two https://www.unfoldedcircle.com/ uses `volume_up` and `volume_down` services. 
I assume https://allonis.com/products/smartremote/smartremote-intro also based on the volume up/down buttons on the remote, but I don't own one so can't verify.

I guess it boils down to all user interfaces where there are buttons for volume up/down. 
Personally I have an automation for an old Hue Dimmerswitch configured to control the media_player on the attic (on/off and brightness buttons for volume) as it is more convenient to have around than using an app.
I have also seen community projects with touch screens that have volume up/down buttons.

While these examples are not part of the Home Assistant project itself, they do use the volume up/down functionality provided by Home Assistant.

**Reply by emontnemery** [reply]
> The `+` and `-` buttons are displayed in the media player dialog. I use them sometimes.

What media player dialog is this? Can you paste a screenshot please?

> [HomeKit Bridge](https://www.home-assistant.io/integrations/homekit/) controls the volume of the media player from iOS Remote Widget through `volume_up` / `volume_down` services.
> Some voice assistants also use volume control through `volume_up` / `volume_down` services, if they receive a command like «Turn up the volume on TV» or «Make TV louder».

For these cases, in particular interacting with voice assistants, I doubt the additional delay of syncing the volume first will make any difference at all?

**Reply by emontnemery** [reply]
@mvdwetering Sure, there are many devices which support the possibility to set an absolute volume and increasing / decreasing the volume. That does not mean Home Assistant's model of media players must use the actual device's functionality for increasing or decreasing the volume.

**Reply by Drafteed** [reply]
> What media player dialog is this? Can you paste a screenshot please?

<img src="https://github.com/home-assistant/architecture/assets/11841379/fb14f3d1-7611-45df-a755-9a6eb1e509d9" width="400">

**Reply by Drafteed** [reply]
> For these cases, in particular interacting with voice assistants, I doubt the additional delay of syncing the volume first will make any difference at all?

@mvdwetering have an excellent example when the delay is critical. 

In addition, frequent call `async_update` (e.g. when we press volume down several times in a short time) can cause failures in the API's which have some "rate limits". In the Bravia TV integration this can cause some troubles: https://github.com/home-assistant/core/issues/91781#issuecomment-1532151925

**Reply by emontnemery** [reply]
Thanks for the screenshot @Drafteed, there's a mismatch between frontend - which shows the `+` and `-` buttons only if `VOLUME_STEP` is set - and core which supports the volume up/down services if either of `VOLUME_STEP` or `VOLUME_SET` are set.

**Reply by emontnemery** [reply]
I'm not sure we can constrain the design by a TV needing 15s for each http request. That's already unacceptably slow, so spamming the `+` and `-` buttons will already not work well although it would of course be even worse if the volume is synced before and/or after the up/down calls.

**Reply by emontnemery** [reply]
Very off-topic for this architecture discussion, but this PR makes sure the `+` and `-` buttons are always shown: https://github.com/home-assistant/frontend/pull/19111

**Comment by Drafteed** [comment]
I think we should leave `volume_up` / `volume_down` services for integrations that can implement relative volume control. We can also extend them to support `volume_step`.

**Reply by emontnemery** [reply]
In that case, how do you suggest to implement a max volume setting then? This is something that has been asked for many times, and many integrations have their own solutions for it.

We could choose to always allow calls to `volume_up` / `volume_down`, but I'm not sure that's acceptable.

If `volume_up` / `volume_down` are suppressed if we expect the call to cause the `max_volume` to be exceeded, that will again only work if the volume has not been changed locally for polled media players.

**Reply by OnFreund** [reply]
If we assume that the volume could be updated outside of HA, what's the value of having max volume enforced only on HA services anyway? Enforcing max volume would require reacting to volume changes on the physical device, and if we do that, we don't need to change the implementation of the services.

**Reply by emontnemery** [reply]
The point of this proposal is to limit what the user can do from Home Assistant. It's obviously not in our power to prevent someone from walking up to the receiver and turning the volume knob to its maximum setting or using another software application to do the same.

**Reply by emontnemery** [reply]
@Drafteed after a lot of discussions, the core team came to the conclusion this may indeed be the least bad option. It's suggested in a new discussion here: https://github.com/home-assistant/architecture/discussions/1050

**Comment by mvdwetering** [comment]
> @mvdwetering Sure, there are many devices which support the possibility to set an absolute volume and increasing / decreasing the volume. That does not mean Home Assistant's model of media players must use the actual device's functionality for increasing or decreasing the volume.

True, but the using the devices relative volume commands always works, even when volume in HA is not (yet) in sync. It will change the volume compared to what I am hearing.

Changing it to an implementation that requires volume in HA to be updated makes it depend on timing of a push update or polling interval. Polling interval delay could be reduced by forcing an immediate update, but that still adds roundtrip time (this could be worked around mostly by making volume updates optimistic)

For me it feels like a step back from the ability to use relative commands offered by the device.

**Comment by emontnemery** [comment]
@Drafteed @mvdwetering @OnFreund there's a new discussion replacing this one: https://github.com/home-assistant/architecture/discussions/1050

**Comment by emontnemery** [comment, marked as answer]
I'm closing this in favor of https://github.com/home-assistant/architecture/discussions/1050

**Comment by balloob** [comment]
Sadly, we never got to a working implementation in Home Assistant (#1050 stalled too), and I think that this is an important feature to have. So I want to refresh the discussion.

- Here is an analysis of all media player entities in Home Assistant and their volumes. There are 77 integrations that provide media player entities with volume controls: https://gist.github.com/balloob/652c3c1f06f24be6899ae49f04e33ba4#compute-current--delta-using-set_volume-override-could-potentially-be-removed
  - Fix for `dunehd` not putting the volume supported features in https://github.com/home-assistant/core/pull/164343
  - Branch removing unnecessary overrides for volume up/down in HA https://github.com/home-assistant/core/compare/claude/fix-volume-overrides-8dboH

The edge cases that stalled this proposal is around integrations natively implementing volume up/down while also having a max volume set. In that case, we cannot know the exact current volume.

That's why I propose it to work like this:

- We support max volume if an entity supports `VOLUME_SET`.
- An entity with a max volume set, will always use our fallback volume up/down methods, which respect max volume
- An entity with no max volume and overridden volume up/down methods, will use the overridden methods.

This results in the entity acting differently when a max volume is set, but that is the point?

**Reply by balloob** [reply]
This is how the proposed implementation would look: https://github.com/home-assistant/core/compare/dev...claude/media-player-volume-registry-4eATp
